connector/connect/src/main/protobuf/spark/connect/types.proto

```diff
@@ -55,134 +57,135 @@ message DataType {
   }
 
-  enum Nullability {
-    NULLABILITY_UNSPECIFIED = 0;
-    NULLABILITY_NULLABLE = 1;
-    NULLABILITY_REQUIRED = 2;
-  }
-
   message Boolean {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message I8 {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message I16 {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message I32 {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message I64 {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message FP32 {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message FP64 {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message String {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message Binary {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message Timestamp {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message Date {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message Time {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message TimestampTZ {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message IntervalYear {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message IntervalDay {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   message UUID {
     uint32 type_variation_reference = 1;
-    Nullability nullability = 2;
+    bool nullability = 2;
   }
 
   // Start compound types.
   message FixedChar {
     int32 length = 1;
     uint32 type_variation_reference = 2;
-    Nullability nullability = 3;
+    bool nullability = 3;
   }
 
   message VarChar {
     int32 length = 1;
     uint32 type_variation_reference = 2;
-    Nullability nullability = 3;
+    bool nullability = 3;
   }
 
   message FixedBinary {
     int32 length = 1;
     uint32 type_variation_reference = 2;
-    Nullability nullability = 3;
+    bool nullability = 3;
   }
 
   message Decimal {
     int32 scale = 1;
     int32 precision = 2;
     uint32 type_variation_reference = 3;
-    Nullability nullability = 4;
+    bool nullability = 4;
+  }
+
+  message StructField {
+    DataType type = 1;
+    string name = 2;
+    bool nullability = 3;
```

Comment:
The current data model assumes that the type itself expresses its nullability, so why do we need nullability here?

I am not saying that the current approach of the datatype having nullability is great. I would also be good if we only set nullability in the StructField message.

Reply:
For a pure type system design, for composable type in which there are sub-types, both parent and children can have nullability.  Use StructType as an example:

1.  Struct(nullability=true, structFields=[(name="a", nullability=true)]):
val x = Struct("a"=NULL) // **valid**
val y: Struct(nullability=true, structFields=[(name="a", nullability=true)] = NULL // **valid**

2. Struct(nullability=true, structFields=[(name="a", nullability=false)]):
val x = Struct("a"=NULL) // **invalid**
val y: Struct(nullability=true, structFields=[(name="a", nullability=true)] = NULL // **valid**

3. Struct(nullability=false, structFields=[(name="a", nullability=false)]):
val x = Struct("a"=NULL) // **invalid**
val y: Struct(nullability=true, structFields=[(name="a", nullability=true)] = NULL // **invalid**


This is very general. I am thinking you are thinking a data model that is a subset of it?

Reply:
This is different from Catalyst. The catalyst data model is:
1. most data types are just enums: string, int, float, etc.
2. some data types have parameters: decimal(3, 1), varchar(10), etc.
3. there are 3 collection types: struct, array, map. These 3 types have nullability: struct field nullability, array element nullability, map value nullability.

Why is proto data type different?

Reply:
I believe the connect proto type system is a bit different on the container type. However I just refactor code here and this is existing code. We don't have tests for types neither.

How about let me follow on this topic later after having more discussions with people. I can add more tests for nullability at that time.